### PR TITLE
Add TS0601 Soil Moisture Sensor

### PR DESCRIPTION
**Pull request by dgaust**
Fixes: #2093

ZHA Quirk to support the Tuya based Soil Moisture sensor. Based off the quirks ts0601_sensor.py and the zigbee-herdsman converter

https://github.com/zigpy/zha-device-handlers/issues/2093

Temperature value is multiplied by 100 to get degrees centigrade Humidty is based on the standard TuyaRelativeHumidity value Battery is multiplied by two to get battery % value

### CODE REVIEW

**Comment by javicalle** [comment]
Thanks for your contribution and **good job** despite 'not 100% sure what I'm doing' 😉 

A few comments:

Edit your first post and put a `Fixes: #2093`. This way, issue will be closed once merged the PR.

I'm not sure if we need to create a new file for this devices. Until now is just another temp/hum sensor.
I would suggest to put the new code in the existing `ts0601_sensor.py`:
* `class SoilManufCluster(TuyaMCUCluster):`  <-- renamed from yours `TemperatureHumidityManufCluster`
* `class TuyaSoilSensor(CustomDevice):`

HA have specific definitions for 'moisture' entities. I will try to check if we can follow the implementations.

It seems that device is reporting also the temp UOM (probably also allows to set it). That would be a improvement, but can go in another PR.

**Comment by javicalle** [comment]
It seems that most of the ZHA magic is already implemented. I would suggest to implement the device like:
* create a new `TuyaSoilCluster`
```python
class TuyaSoilMoisture(SoilMoisture, TuyaLocalCluster):
    """Tuya local SoilMoisture cluster."""

```

You will need to import the `SoilMoisture` class:
```python
from zigpy.zcl.clusters.measurement import (
    RelativeHumidity,
    SoilMoisture,
    TemperatureMeasurement,
)
```

* use the new class in your device (`dp_to_attribute` and `replacement`)

I believe that HA will create the new entity. And probably will be pretty similar to the old one but IMHO 'this is the way'.

**Comment by dgaust** [comment]
> I was thinking about the same, but keep merged for now. Let's see how it grows first.

Ok, I've made those changes to the file, bringing the standalone file back into the sensor.py version. Have also added the soil moisture class rather than using humidity. It did create a new sensor in Home Assistant and all seems to be working fine

image

**Comment by javicalle** [comment]
You have the imports duplicated. Keep the 'new' format.

The temp OUM will not be an easy fix (ZHA assumes that temp units will be Celsius always). I believe that we can go 'as is'.

**Comment by dgaust** [comment]
> You have the imports duplicated. Keep the 'new' format.
> 
> The temp OUM will not be an easy fix (ZHA assumes that temp units will be Celsius always). I believe that we can go 'as is'.

Yep, I'd noticed that once it failed and have made the adjustments.

**Comment by javicalle** [comment]
You can check the isort report for more info:
https://github.com/zigpy/zha-device-handlers/actions/runs/3934353893/jobs/6729028066

The check complains about the extra break lines (the `-` lines). Delete it.

**Comment by dgaust** [comment]
> You can check the isort report for more info: https://github.com/zigpy/zha-device-handlers/actions/runs/3934353893/jobs/6729028066
> 
> The check complains about the extra break lines (the `-` lines). Delete it.

Thansk, I didn't know how to interpret what the issue was :D

All fixed and tests passed.

File: zhaquirks/tuya/ts0601_sensor.py [hunk 1]
```diff
@@ -24,6 +28,20 @@ class TuyaTemperatureMeasurement(TemperatureMeasurement, TuyaLocalCluster):
...
+        """Apply a correction factor to value."""
+        if attr_name == "measured_value":
+            value = value * (
+                self.endpoint.device.RH_MULTIPLIER
+                if hasattr(self.endpoint.device, "RH_MULTIPLIER")
+                else 100
+            )
+        return super().update_attribute(attr_name, value)
+
+
```
Comment (javicalle) [inline zhaquirks/tuya/ts0601_sensor.py:44]: Remove the `update_attribute` method here. The `RH_MULTIPLIER` sates for RelativeHumidity. Keep just the class definition like:
```python
class TuyaSoilMoisture(SoilMoisture, TuyaLocalCluster):
    """Tuya local SoilMoisture cluster."""


```
(2 break lines before next class `TuyaRelativeHumidity`)

File: zhaquirks/tuya/ts0601_sensor.py [hunk 2]
```diff
@@ -207,3 +224,77 @@ class TuyaTempHumiditySensorVar03(CustomDevice):
...
+        5: DPToAttributeMapping(
+            TuyaTemperatureMeasurement.ep_attribute,
+            "measured_value",
+            dp_type=TuyaDPType.VALUE,
+            converter=lambda x: x * 100,
+        ),
+        3: DPToAttributeMapping(
+            TuyaSoilMoisture.ep_attribute,
+            "measured_value",
+            dp_type=TuyaDPType.VALUE,
```
Comment (javicalle) [inline zhaquirks/tuya/ts0601_sensor.py:242]: If removed from method, we need to add here the value conversion like:
```python
            dp_type=TuyaDPType.VALUE,
            converter=lambda x: x * 100,
```

**Review by javicalle — CHANGES_REQUESTED** [review]
A little fix to remove the `update_attribute` method.
If needed in a future we can go back to it.

**Comment by javicalle** [comment]
With the last changes the PR will be fine to me.
But can you check yhe device logs for any warning about unsupported handlers or similar. There must be a DP=9 for the UOM

**Review by javicalle — APPROVED** [review]
LGTM

**Pull request merged**